test/core.jl

```diff
@@ -3140,6 +3140,38 @@ end
 @test_throws TypeError MyType8010([3.0;4.0])
 @test_throws TypeError MyType8010_ghost([3.0;4.0])
 
+module TestNewTypeError
+using Base.Test
+
+struct A
+end
+struct B
+    a::A
+end
+@eval function f1()
+    # Emitting this direction is not recommended but it can come from `convert` that does not
+    # return the correct type.
+    $(Expr(:new, B, 1))
+end
+@eval function f2()
+    a = $(Expr(:new, B, 1))
+    a = a
+    return nothing
+end
+@generated function f3()
+    quote
+        # This is currently broken since the expression is removed by the frontend.
```

Comment:
This is a valid AST that can be produced by the parser. I just use this syntax to avoid having to define a convert.

```julia
julia> struct A
       end

julia> struct B
           a::A
           function B(x)
               new(x)
               return
           end
       end

julia> Base.convert(::Type{A}, x::Int) = x

julia> B(1)

julia> @code_lowered B(1)
CodeInfo(:(begin
        nothing
        #= line 5 =#
        return
    end))
```